probe_src/probe_py/analysis.py

```diff
@@ -59,59 +70,44 @@ def last(pid: int, exid: int, tid: int) -> Node:
             exec_edges.append((node, first(*target)))
         elif isinstance(op, WaitOp) and op.ferrno == 0 and op.ret > 0:
             # Always wait for main thread of the last exec epoch
             if op.ferrno == 0:
-                target = (op.ret, last_exec_epoch.get(op.ret, 0), op.ret)
+                target = (op.ret, -1, last_exec_epoch.get(op.ret, 0), op.ret)
                 fork_join_edges.append((last(*target), node))
         elif isinstance(op, ExecOp):
             # Exec brings same pid, incremented exid, and main thread
-            target = pid, exid + 1, 0
+            target = pid, _time, exid + 1, 0
             fork_join_edges.append((node, first(*target)))
-
-    # # Make the main thread exit at the same time as each thread
-    for pid, process in process_tree_prov_log.processes.items():
+            
+    # Make the main thread exit at the same time as each thread
+    for (pid, _time), process in process_tree_prov_log.processes.items():
         for exec_epoch_no, exec_epoch in process.exec_epochs.items():
             for tid, thread in exec_epoch.threads.items():
                 if tid != 0:
-                    fork_join_edges.append((last(pid, exec_epoch_no, tid), last(pid, exec_epoch_no, pid)))
-
-    def node_to_label(node: Node) -> str:
-        inner_label = node[3] if isinstance(node[3], str) else node[3].__class__.__name__
-        return f"{node[0]} {node[1]} {node[2]} {inner_label}"
-
-    def node_to_id(node: Node) -> str:
-        return f"\"node_{node[0]}_{node[1]}_{node[2]}_{id(node[3])}\""
-
-    return "\n".join([
-        "strict digraph {",
-        *[
-            f"  {node_to_id(node)} [label=\"{node_to_label(node)}\"];"
-            for node in nodes
-        ],
-        *[
-            f"  {node_to_id(node0)} -> {node_to_id(node1)} [color=\"green\"];"
-            for node0, node1 in program_order_edges
-        ],
-        *[
-            f"  {node_to_id(node0)} -> {node_to_id(node1)} [color=\"yellow\"];"
-            for node0, node1 in exec_edges
-        ],
-        *[
-            f"  {node_to_id(node0)} -> {node_to_id(node1)} [color=\"red\"];"
-            for node0, node1 in fork_join_edges
-        ],
-        "}",
-    ])
+                    fork_join_edges.append((last(pid, _time, exec_epoch_no, tid), last(pid, _time, exec_epoch_no, 0)))
+
+    process_graph = nx.DiGraph()
+    
+    for node in nodes:
+        # op = [pid, _time, exec_id, tid, op_index]
+        process_graph.add_node(node[4], op = [node[0], node[6], node[1],node[2],node[5]])
+    
+    for node0, node1 in program_order_edges:
+        process_graph.add_edge(node0[4], node1[4], color = 'green')
+    
+    for node0, node1 in exec_edges:
+        process_graph.add_edge(node0[4], node1[4], color = 'yellow')
+
+    for node0, node1 in fork_join_edges:
+        process_graph.add_edge(node0[4], node1[4], color='red')
+
+    return nx.drawing.nx_pydot.write_dot(process_graph,"./processgraph")
```

Comment:
We should split this function into two: One that converts `ProvLog` to `nx.DiGraph` (no colors), and another that turns `nx.DiGraph` to `str` (pydot string with graphics).

To turn the output of `write_dot` to a string, we should try using a temporary in-memory buffer
```
buffr = io.StringIO
nx.drawing.nx_pydot.write_dot(process_graph, buffr)
return buffr.getvalue()
```

Let me know if this approach doesn't work.

Reply:
@charmoniumQ this approach does not work, because `write_dot` function takes path as an input parameter no `StringIO`. Can we do something like this instead? I tried this.
```
pydot_graph = nx.drawing.nx_pydot.to_pydot(process_graph)
dot_string = pydot_graph.to_string()
buffer = io.StringIO()
buffer.write(dot_string)
return buffer.getvalue()
```

Reply:
Ah, we just need a string, so in that case, just

```
pydot_graph = nx.drawing.nx_pydot.to_pydot(process_graph)
return pydot_graph.to_string()
```

:)

Reply:
@charmoniumQ So the function `provlog_to_digraph` will have all the edges but we won't assign colors to the edges, and the second function `digraph_to_buffer` will add colors to the edges and return the pydot string using `buffer`. Am i on point?